service/wechat/wechat.go

```diff
@@ -0,0 +1,144 @@
+package wechat
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/silenceper/wechat/v2"
+	"github.com/silenceper/wechat/v2/cache"
+	"github.com/silenceper/wechat/v2/officialaccount/config"
+	"github.com/silenceper/wechat/v2/officialaccount/message"
+	"github.com/silenceper/wechat/v2/util"
+)
+
+// Config is the Service configuration.
+type Config struct {
+	AppID          string
+	AppSecret      string
+	Token          string
+	EncodingAESKey string
+}
+
+// wechatMessageManager abstracts go-wechat's message.Manager for writing unit tests
+type wechatMessageManager interface {
+	Send(msg *message.CustomerMessage) error
+}
+
+// Service encapsulates the WeChat client along with internal state for storing users.
+type Service struct {
+	config         Config
+	messageManager wechatMessageManager
+	userIDs        []string
+}
+
+// New returns a new instance of a WeChat notification service.
+func New(cfg Config) *Service {
+	wc := wechat.NewWechat()
+	wcCfg := &config.Config{
+		AppID:          cfg.AppID,
+		AppSecret:      cfg.AppSecret,
+		Token:          cfg.Token,
+		EncodingAESKey: cfg.EncodingAESKey,
+		Cache:          cache.NewMemory(),
+	}
+
+	oa := wc.GetOfficialAccount(wcCfg)
+
+	return &Service{
+		config:         cfg,
+		messageManager: oa.GetCustomerMessageManager(),
+	}
+}
+
+// WaitForOneOffVerification waits for the verification call from the WeChat backend.
+//
+// Should be running when (re-)applying settings in wechat configuration.
+//
+// Set devMode to true when using the sandbox.
+//
+// See https://developers.weixin.qq.com/doc/offiaccount/en/Basic_Information/Access_Overview.html
+func (s *Service) WaitForOneOffVerification(
```

Comment:
Any reason why we don't use [this](https://github.com/silenceper/wechat/blob/release-2.0/officialaccount/server/server.go) internally? We could store the in *wechat.New* created account in the service struct and then use the accounts internal server here instead of all this boilerplate code.

Don't get me wrong, really appreciate your work but I'm just looking for a simpler solution here. Since you surely have a deeper understanding of the WeChat library maybe you can explain me why we're not using [the accounts server](https://github.com/silenceper/wechat/blob/release-2.0/officialaccount/server/server.go) here. Very likely that I'm just overlooking something here.

I will not review the code in this function until this is solved.

Reply:
@nikoksr I was puzzled with this as well. This is the argumentation:

Wechat API is geared towards two-way communication and requires a verification on the existence of the other party. The verification happens only once per application-host address combination. If the host changes, it needs to happen again, as Wechat needs to know where to sends messages to. In any case, the verification is a **manually-invoked** (user just clicks a button on the portal to make the http call) and **non-recurring** process (wechat will not attempt to do this on its own). 

In addition, notify is not two-way communication and as such any change in the host address will not break anything. Practically, this means that verification for notify is an **once-in-lifetime** process for simply activating the application (ID). In other words, there is no reason to have a running [the accounts server](https://github.com/silenceper/wechat/blob/release-2.0/officialaccount/server/server.go). We only need this for the once-in-lifetime verification.

Profound conclusion would be to use the accounts server for the verification and shut it down after. That would have saved as the trouble of coding the `WaitForOneOffVerification`. Unfortunately, as you can see [here](https://github.com/silenceper/wechat/blob/5472ac979b097d2841fda5792142311c1c746525/officialaccount/server/server.go#L56), we are not given a callback on the verification message. That means that we cannot know whether the verification has happened, in order to shutdown server or allow notifications to be sent out. 

As such, I created this helper function that already utilises the accounts server internal signature verification function but doesn't re-use the whole server. If you prefer, I can remove the `WaitForOneOffVerification` and simply add instructions for running the accounts server in order to perform the verification. One downside to this is that the accounts server has no graceful shutdown.

Reply:
@gtourkas okay, this is indeed confusing. Thank you for your detailed explanation and research. I think you made the right decision here. We will stick with your implementation. Thanks for your efforts, I will now review the function.